utils/mil_tools/nodes/online_bagger.py

```diff
@@ -319,78 +313,146 @@ def set_bag_directory(self, filename=''):
+                else:
+                    index = self.get_time_index(topic, requested_seconds)
+                    total_messages += len(self.topic_messages[topic][index:])
+                    bag_topics[topic] = index
+            if total_messages == 0:
+                result.success = False
+                result.status = 'no messages'
+                self._action_server.set_aborted(result)
+                self.streaming = True
+                bag.close()
+                return
+            self._action_server.publish_feedback(feedback)
+            msg_inc = 0
+            for topic, index in bag_topics.iteritems():
+                for msgs in self.topic_messages[topic][index:]:
+                    bag.write(topic, msgs[1], t=msgs[0])
+                    if msg_inc % 50 == 0:  # send feedback every 50 messages
+                        feedback.progress = float(msg_inc) / total_messages
+                        self._action_server.publish_feedback(feedback)
+                    msg_inc += 1
+                    # empty deque when done writing to bag
+                self.topic_messages[topic].clear()
+            feedback.progress = 1.0
+            self._action_server.publish_feedback(feedback)
+            bag.close()
+        except Exception as e:
+            result.success = False
+            result.status = 'Exception while writing bag: ' + str(e)
+            self._action_server.set_aborted(result)
+            self.streaming = True
+            bag.close()
+            return
+        rospy.loginfo('Bag written to {}'.format(result.filename))
+        result.success = True
+        self._action_server.set_succeeded(result)
         self.streaming = True
-        return self.bagger_status
 
 
+class OnlineBaggerClient(object):
+    '''
+    Wrapper to run an action client connecting to online bagger
+    and triggering a write with the specified topics, time, and
+    filename. Uses tqdm to display a progress bar as the write
+    occurs.
+    '''
+    def __init__(self, name='', topics='', time=0.0):
+        self.client = SimpleActionClient('/online_bagger/bag', BagAction)
+        self.goal = BagGoal(bag_name=name, topics=topics, bag_time=time)
+        self.result = None
+        self.last_progress = 0
+
+    def _done_cb(self, status, result):
+        self.result = (status, result)
+
+    def _feedback_cb(self, feedback):
+        iterations = int((feedback.progress - self.last_progress) / 0.01)
+        if iterations >= 1:
+            rospy.set_param('/test', feedback.progress)
+            self.bar.update(iterations)
+            self.bar.refresh()
+            self.last_progress = feedback.progress
+
+    def bag(self, timeout=rospy.Duration(0)):
+        self.client.wait_for_server()
+        self.result = None
+        self.last_progress = 0.0
+        self.bar = tqdm(desc='Writing bag', unit=' percent', total=100,
+                        bar_format='{desc} {bar} {percentage:3.0f}%')
```

Comment:
Looks where this code is. This is only run when online bagger is run with ```-c```, which runs an action client.